lib/WUI/wui.cpp

```diff
@@ -0,0 +1,620 @@
+        HAL_StatusTypeDef status = HAL_RNG_GenerateRandomNumber(&hrng, &random);
+        if (HAL_OK == status) {
+            api_key[i++] = charset[random % charset_length];
+        }
+    }
+    api_key[i] = 0;
+    return api_key;
+}
+
+void wui_store_api_key(char *api_key, uint32_t length) {
+    variant8_t *p_prusa_link_api_key = &prusa_link_api_key;
+    variant8_done(&p_prusa_link_api_key);
+    prusa_link_api_key = variant8_init(VARIANT8_PCHAR, length, api_key);
+    eeprom_set_var(EEVAR_PL_API_KEY, prusa_link_api_key);
+}
+
+namespace {
+
+void prusalink_api_key_init(void) {
+    prusa_link_api_key = eeprom_get_var(EEVAR_PL_API_KEY);
+    if (!strcmp(variant8_get_pch(prusa_link_api_key), "")) {
+        char api_key[PL_API_KEY_SIZE] = { 0 };
+        wui_generate_api_key(api_key, PL_API_KEY_SIZE);
+        wui_store_api_key(api_key, PL_API_KEY_SIZE);
+    }
+}
+
+// This is the top-level manager of network settings and interfaces.
+//
+// It is responsible for bringing up all the interfaces with the right settings
+// and reloading the settings as needed. It doesn't do much functionality on
+// the network itself (it does DHCP, which is part of its core
+// responsibilities, it starts the http server and ntp client ‒ starting them
+// is probably part of the management responsibilities, but some kind of hooks
+// mechanism might be better in the future).
+//
+// Technicalities:
+//
+// * In practice, the thing needs to be a singleton. We need that because
+//   interrupts need to be sent somewhere, the rest of the printer needs to have
+//   something to contact, etc. But the logic is not singletonish inside.
+// * The external API is legacy ‒ taken from the original implementation. This
+//   is transitional measure and some kind of newer & nicer interface will be
+//   done later on. Maybe as some kind of facade on top of it.
+// * This lives in its own thread/task. It handles both the reconfiguration and
+//   reading packets from both our interfaces. Nevertheless, sending of packets
+//   happens from whatever task they happen to be created. Similarly, accessing
+//   the current configuration can happen from whatever.
+// * We are lazy and do any kind of reconfiguration by shutting everything
+//   down, reading new state from EEPROM and then bringing it up again. When
+//   someone wants to change something, it is first written into EEPROM, then a
+//   reload is signalled and happens in our thread.
+// * While writing configuration is synchronized through the write to the
+//   EEPROM, reading of the configuration needs to happen inside a lock. This is
+//   not yet written, but we at least have a well defined place where it goes
+//   inside this class.
+class NetworkState {
+public:
+    enum NetworkAction {
+        CoreInitDone = 1 << 0,
+        Reconfigure = 1 << 1,
+        EthInitDone = 1 << 2,
+        EthData = 1 << 3,
+        EspInitDone = 1 << 4,
+        EspData = 1 << 5,
+        TriggerNtp = 1 << 6,
+    };
```

Comment:
What's `BIT(2)`? Are you suggesting I should create a macro `BIT(OFFSET) 1 << OFFSET` or something like that?

The struct won't work (well, not without a lot of tedious boilerplate), because it doesn't have working bit-manipulating operations. So one would have to go and define all the `&` and `|` operators for it and constants… it would be tedious to work with. This way it's not perfect, but it's used locally (inside single cpp file), so ease of creation and usage won.